presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeToTimestampCast.java

```diff
@@ -56,23 +57,36 @@ public static LongTimestamp castToLong(
             ConnectorSession session,
             @SqlType("time(sourcePrecision)") long time)
     {
-        long epochMicros = cast(sourcePrecision, targetPrecision, session, time);
-        return new LongTimestamp(epochMicros, (int) (time % PICOSECONDS_PER_MICROSECOND));
+        long epochSeconds = getEpochSeconds(session, time);
+        long picoFraction = getPicoFraction(sourcePrecision, targetPrecision, time);
+        long epochMicros = computeEpochMicros(epochSeconds, picoFraction);
+
+        int picosOfMicro = (int) (picoFraction % PICOSECONDS_PER_MICROSECOND);
+        return new LongTimestamp(epochMicros, picosOfMicro);
     }
 
-    private static long cast(long sourcePrecision, long targetPrecision, ConnectorSession session, long time)
+    private static long getEpochSeconds(ConnectorSession session, long time)
     {
         // TODO: consider using something more efficient than LocalDate.ofInstant() to compute epochDay
         long epochDay = LocalDate.ofInstant(session.getStart(), session.getTimeZoneKey().getZoneId())
                 .toEpochDay();
 
-        long epochSecond = multiplyExact(epochDay, SECONDS_PER_DAY) + time / PICOSECONDS_PER_SECOND;
+        return multiplyExact(epochDay, SECONDS_PER_DAY) + time / PICOSECONDS_PER_SECOND;
+    }
+
+    private static long getPicoFraction(long sourcePrecision, long targetPrecision, long time)
+    {
         long picoFraction = time % PICOSECONDS_PER_SECOND;
         if (sourcePrecision > targetPrecision) {
-            picoFraction = round(picoFraction, (int) (TimeType.MAX_PRECISION - targetPrecision));
+            picoFraction = round(picoFraction, (int) (12 - targetPrecision));
         }
 
-        long microFraction = rescale(picoFraction, TimeType.MAX_PRECISION, 6);
-        return multiplyExact(epochSecond, MICROSECONDS_PER_SECOND) + microFraction;
+        return picoFraction;
+    }
+
+    private static long computeEpochMicros(long epochSeconds, long picoFraction)
+    {
+        long microFraction = rescale(picoFraction, 12, 6);
```

Comment:
Does `rescale` round or truncate?
i checked it truncates, but i'd write this simply as
```
// picoFraction is already rounded to whole micros
long microFraction = picoFraction / PICOSECONDS_PER_MICROSECOND;
```